src/stats/cdf.rs

```diff
@@ -0,0 +1,209 @@
+// except according to those terms.
+
+//! Support for discrete probability distributions in terms of cumulative distribution
+//! functions (CDF).
+
+use std::f64;
+use std::iter;
+use std::slice;
+
+use num::traits::{cast, NumCast};
+use itertools::Itertools;
+
+use stats::logprobs::{self, LogProb};
+
+
+/// Implementation of a cumulative distribution function.
+#[derive(Debug, Clone)]
+pub struct CDF<T: PartialOrd> {
+    inner: Vec<(T, LogProb)>
+}
+
+
+impl<T: PartialOrd> CDF<T> {
+    /// Create CDF from given probability mass function (PMF). The PMF may contain duplicate values
+    /// the probabilities of which are summed during generation of the CDF.
+    ///
+    /// # Arguments
+    ///
+    /// * `pmf` - the PMF as a vector of value/probability pairs
+    pub fn from_pmf(mut entries: Vec<(T, LogProb)>) -> Self {
+        entries.sort_by(|&(ref a, _), &(ref b, _)| a.partial_cmp(b).unwrap());
+        let mut inner: Vec<(T, LogProb)> = Vec::new();
+        for mut e in entries.into_iter() {
+            let p = logprobs::add(inner.last().map_or(f64::NEG_INFINITY, |e| e.1), e.1);
+            if !inner.is_empty() && inner.last().unwrap().0 == e.0 {
+                inner.last_mut().unwrap().1 = p;
+            }
+            else {
+                e.1 = p;
+                inner.push(e);
+            }
+        }
+        let mut cdf = CDF {
+            inner: inner
+        };
+
+        if relative_eq!(cdf.total_prob(), 0.0) && cdf.total_prob() > 0.0 {
+            cdf.inner.last_mut().unwrap().1 = 0.0;
+        }
+
+        cdf
+    }
+
+    /// Create CDF from iterator. This can be used to replace the values of a CDF.
+    pub fn from_cdf<I: Iterator<Item = (T, LogProb)>>(entries: I) -> Self {
+        CDF { inner: entries.collect_vec() }
+    }
+
+    /// Reduce CDF by omitting values with zero probability.
+    pub fn reduce(self) -> Self {
+        let mut inner = Vec::new();
+        let mut last = f64::NEG_INFINITY;
+        for e in self.inner.into_iter() {
+            if last != e.1 {
+                last = e.1;
+                inner.push(e);
+            }
+        }
+        CDF { inner: inner }
+    }
+
+    /// Downsample CDF to n entries.
+    pub fn sample(mut self, n: usize) -> Self {
+        if self.inner.len() <= n {
+            self
+        }
+        else {
+            let s = self.inner.len() / (n - 1);
+            let last = self.inner.pop().unwrap();
+            let mut inner = self.inner.into_iter().step(s).collect_vec();
+            inner.push(last);
+            CDF {
+                inner: inner
+            }
+        }
+    }
+
+    /// Provide iterator.
+    pub fn iter(&self) -> slice::Iter<(T, f64)>{
+        self.inner.iter()
+    }
+
+    /// Iterator over corresponding PMF.
+    pub fn iter_pmf<'a>(&'a self) -> CDFPMFIter<'a, T> {
+        fn cdf_to_pmf<'a, G>(last_prob: &mut LogProb, e: &'a (G, LogProb)) -> Option<(&'a G, LogProb)> {
```

Comment:
Nit: can you use a name other than `e`? I really have no idea what this is here.